### DISCUSSION TITLE
Livewire 3 Alpine doesn't work with Laravel Mix

### DISCUSSION THREAD

**Discussion by adriansio** [livewire/livewire#6844] [1. Help]
I am trying to upgrade to Livewire 3 and I can't get Alpine to work with the bundling method (I am using Laravel Mix, I can't use Vite).

This is my `webpack.mix.js`:

```
const mix = require('laravel-mix');

mix.js('resources/js/app.js', 'public/js');
```

This is my `app.js`:

```
import {Alpine, Livewire} from '../../vendor/livewire/livewire/dist/livewire.esm';
import Clipboard from '@ryangjchandler/alpine-clipboard'

Alpine.plugin(Clipboard)

Livewire.start()
```

And this is my `welcome.blade.php`:

```
<html>

<head>
    @livewireStyles

    <script src="{{ mix('js/app.js') }}" defer></script>
</head>

<body>
    <h1 x-data="{ message: 'I ❤️ Alpine' }" x-text="message"></h1>

    @livewireScriptConfig
</body>

</html>
```

After compiling with `npm run dev`, I keep getting this error:

```
Alpine Expression Error: func(...).catch is not a function
Uncaught TypeError: func(...).catch is not a function
```

Is there really no way to use Livewire 3, Alpine, and Laravel Mix with the bundling method?

**Comment by benjibee** [comment]
We're running into the same issue in our project just now. Did you find any good workaround or fix?

I wonder why there aren't more issues about this, is everyone really on Vite?

**Comment by ojsoft** [comment]
I use Laravel mix also. I removed this code from app.js.

In your case just write (No livewire start, no Alpine start, will be done all by livewire inject_assets option):

```
import Clipboard from '@ryangjchandler/alpine-clipboard'
Alpine.plugin(Clipboard)
```

See livewire.php config (maybe publish first)
```
/*
    | By default, Livewire automatically injects its JavaScript and CSS into the
    | <head> and <body> of pages containing Livewire components. By disabling
    | this behavior, you need to use @livewireStyles and @livewireScripts.
    |
    */

    'inject_assets' => true,

```

Also remove @livewireStyles and @livewireScriptConfig and @livewireScripts from your code

This worked at least for me, I had the same problem in the beginning.

ATTENTION: Blut please be aware that alpinejs is then only included if a livewire component is on the page. If this seems to be a bad idea in your case, you have to take care of that.

**Comment by jensolafkoch** [comment]
I was struggling with that as well. My final solution after some hours was to use the standard injection from livewire, meaning no @livewireStyles,@livewireScriptConfig or @livewireScripts directives, no code in app.js, and thus no Alpine at all, because the docs say:

> Because Alpine is bundled with Livewire's JavaScript assets, you must include @livewireScripts on every page you wish to use Alpine. Even if you're not using Livewire on that page.

I then added Alpine 3 again as usual (package.json and app.js), that is: as if I had no Livewire at all ...

```
import Alpine from 'alpinejs'
import persist from '@alpinejs/persist'
import collapse from '@alpinejs/collapse'
import ToastComponent from '../../vendor/usernotnull/tall-toasts/resources/js/tall-toasts'
Alpine.plugin(persist)
Alpine.plugin(collapse)
Alpine.data('ToastComponent', ToastComponent)
window.Alpine = Alpine
Alpine.start()
```

I made a test component and included it in my master layout and tested if I had access to Alpine: I do.

I don't know yet if that may cause other problems, but wanted to share anyway - and hoping for feedback: Do you think that would bne a stable solution for the time being?

UPDATED:

Hm, there is a problem of course. When I render the ToastComponent from above I get errors like

`Alpine Expression Error: $wire is not defined`

so my attempt to "mix" Livewire with integrated Alpine and the separate "external" Alpine package does not work ...

Any better ideas? :-)

**Comment by adriansio** [comment]
I couldn't find any solution because I need some Alpine plugins.
My only solution was to spend countless hours and switch to Vite (still in the process of doing it).

**Reply by jensolafkoch** [reply]
And does that solve the Alpine problem so that the extra plugins work? I need a special plugin which is not part of Livewire/Alpine 3, so switching to Vite would only help if the bundling option does really work with it.

**Reply by lrljoe** [reply]
If you bundle, rather than using injection, then it 100% works, you can add your plugins into your bundle easily.

Using injection, it depends on how the plugin is developed!

**Comment by lrljoe** [comment]
Laravel Shift is your friend to migrate from mix to Vite simply.

If it's taking more than moments to switch bundler approaches, then something isn't quite right!

I tried getting Mix working, just to be helpful, but given the last release was several years ago, it wasn't worth it for my use case, and I used Shift to migrate the impacted projects, taking around 20mins to migrate 6 large projects.

**Reply by jensolafkoch** [reply]
I'm thinking about Vite (I use it in all newer projects) for quite some time, but in this very project I deal with "foreign" legacy code which is a bit chaotic. Lots of package resources are copied manually in Mix from node_modules instead of being integrated tzo the build, often quite old packages, I began to first integrate them into the Mix build but lots of problems. As far as I know Vite doesn't come with the option to just copy resources.

And does Vite really solve the problem with the Alpine plugin problem when adding Livewire via bundling?

**Reply by lrljoe** [reply]
Vite absolutely can just copy files around, or do vendor split, but that's definitely a Vite question more than Livewire!

**Reply by jensolafkoch** [reply]
Okay thanks! So: Bundling Alpine plugins does not work with Mix (which is the maikn problem here) but it does with Vite, and you know from personal experience? (Sorry to ask again, but switching to Vite in this legacy project will be a nightmare anyway because of the many legacy scripts and their weird integration.)

**Reply by lrljoe** [reply]
So for LW3 I use custom Alpine plugins, and I also use the same LW3 core injection methods to add more Alpine methods/data elements.

**Reply by jensolafkoch** [reply]
My notion about Vite's lack was based on the Mix to Vite path described in the Laravel docs which say:

> However, it lacks some features that Laravel Mix supports, such as the ability to copy arbitrary assets into the build that are not referenced directly in your JavaScript application.

So that's no longer the case? That would be great news.

**Comment by jensolafkoch** [comment]
So your integration of Alpine plugins resembles the way recommended by the LW3 docs, like ...

```
import {Livewire, Alpine} from '../../vendor/livewire/livewire/dist/livewire.esm';
import ToastComponent from '../../vendor/usernotnull/tall-toasts/resources/js/tall-toasts'

require('./bootstrap'); // -> popper.js, jquery, axios, csrf-token

Alpine.plugin(ToastComponent)

Livewire.start() //
```

**Reply by lrljoe** [reply]
That looks correct to me!  There are other native apps that you may need.

What else are you trying to include, as there are various ways to get things in

**Reply by jensolafkoch** [reply]
Too much to deal with it here :-), there are about 20 CSS and JS resources copied around by Mix and then individually included in script tags in a delicate order. A nightmare. I will probably leave it alone for the time being and try to duplicate the same workflow with Vite some time.

But for the moment, in the hope that the problem with Mix might be fixed later, I decided to only add LW2 for the time being so that I can use Alpine independently.

**Reply by lrljoe** [reply]
I mean, you can hope!  I've looked, no joy, the core team has looked, no joy.  The last mix commit was 3 years ago, so if you have any customers with even a toddler approach to security, they'll be asking questions!
Mix is, pretty much abandoned, and the problem is that it doesn't natively support newer es versions, and honestly, I can't see the livewire team taking on upgrading and maintaining mixml.

With Vite, you can still use all of the core push etc elements, you can split code up, and only include it on certain pages etc.

I'm struggling with understanding the limitations you're facing to be honest.  I've overseen migrations in .gov and pci compliant environments, and am just not sure about why you feel it'd be problematic

I don't think mix is even supported at all in Laravel 11, I could be wrong tho!

**Reply by jensolafkoch** [reply]
Yes, Vite is clearly the way to go, you are absolutely right! :-)

Well, I mostly work in cleanly set up new projects, so my experience with build processes is somewhat limited (although I do my best :-)). And I missed a couple of years (say 15) in dev between early HTML and CSS stages and current frameworks, so I missed crucial evolutionary steps in between - thus a certain lack of experience especially in bundling ... I only "back" for around five years, and I was quite happy to manage Mix ...

**Reply by lrljoe** [reply]
The discord community is very active and helpful if you get stuck by the way, everyone has been through this at some point!  It's almost entirely text based, so you won't get deafened by anyone either.

And I totally get what you mean, I found the transition daunting having spent years in a non dev capacity, finally figuring out mix, then having to  jump to vite.  It is simple once you "get it", until then, the docs may as well be written in martian.
  I'm definitely not a fan of either vite or mix docs, as they all seem a bit too elitest, and unnecessarily complex, especially when compared to some of the other projects.

**Reply by jensolafkoch** [reply]
Thanks, that's really encouraging! And you have definitely a point here re the docs ... :-)

**Reply by adriansio** [reply]
Vite is great actually, it's very fast. Just need to get the hang of it. The thing that drives me nuts is the fact that I can't find a way to completely disable "vendor chunking". So I am fighting against Vite and using two different Vite configs.

**Reply by lrljoe** [reply]
```
build: {
    rollupOptions: {
        output: {
            manualChunks: () => 'app',
        },
    },
},
```

will disable chunking and put it all into app.js and app.css

You can manually specify which files to chunk separately should you wish to as well.

**Reply by adriansio** [reply]
I can't put anything into app because I have some widgets.js that need to be built separately.

**Comment by slimani-dev** [comment]
this worked for me

*bootstrap.js*
```javascript


if (!window.Alpine) {
    import('../../vendor/livewire/livewire/dist/livewire.esm').then(module => {
        const { Livewire, Alpine } = module
        // Optionally, attach Livewire and Alpine to the window for global access
        window.Livewire = Livewire
        window.Alpine = Alpine
        Livewire.start()
    }).catch(error => {
        console.error('Failed to load Livewire and Alpine:', error)
    })
}

```
